### PR TITLE
Load CDK: Support strongly/weakly-typed mappers

### PR DESCRIPTION
**Pull request by edgao**
closes https://github.com/airbytehq/airbyte-internal-issues/issues/11217

**strongly** recommend reviewing by commit. Overall change looks scary (... and individual commits are still scary), but they hopefully make sense. A lot of the diff is just dealing with the class/function signature changes.

This came out of https://airbytehq-team.slack.com/archives/C03AS1GAQV6/p1734637148661599. Backgroundt:
* Some destinations want to operate on "real" types (java.time.OffsetDateTime, etc) - avro, iceberg, etc.
* Some destinations don't care about types at all - JSONL, CSV
* Some destinations only care about top-level fields' types (typical DB/DW destinations)

Prior to this PR, it was actually kind of tricky to get those java.time.XYZ types without either:
* hard failing on invalid values (i.e. if a source declares a field of type timestamptz, then emits `"potato"`)
* or double-parsing timestamps (i.e. having a mapper to validate the field, and then later re-parsing it to a java.time.XYZ object)

Also, untyped destinations (jsonl/csv) would still null out fields for being the "wrong" type.

After this PR:
* Untyped destinations are pure pass-through
* Fully-typed destinations (avro, parquet, iceberg) can use AirbyteValueDeepCoercingMapper to coerce all types everywhere
* (not implemented, but theoretically possible) root-level-typed destinations (db/dw) can write a mapper to coerce the top-level fields, without recursing into nested objects/arrays

## JsonNode -> AirbyteValue conversion now ignores the schema

In other words, the first pass marshaling into AirbyteValue is basically a one-to-one mapping from JsonNode.type onto AirbyteValue. And then a bunch of changes to make the compiler happy.

## Implement a type-coercing mapper

A couple things happening here:
* date/time/timestamp AirbyteValue now have actual java.time.XYZ values, instead of String
    * also, add time/timestamp types for with/without timezone
* AirbyteValueIdentityMapper no longer does any casting - i.e. it always passes through the data unchanged
* add AirbyteValueDeepCoercingMapper for destinations that require the values' types to match the schema
* and again, compiler fixes

## Fixing behavior for non-coerced destinations

* change the json+csv data dumpers to not do any type coercion - i.e. they're also doing a naive schema-ignorant conversion to AirbyteValue
* implement UncoercedExpectedRecordMapper, to make tests pass again (this just downcasts time/etc. values to StringValue, to match the new dumper behavior)
* a few related bugfixes

## Fixing behavior for avro/parquet

* tweak the mapper pipelines to use the new AirbyteValueDeepCoercingMapper
    * moves the MergeUnions mapper to the top of the pipeline - afaict this works fine, and is basically a requirement for the coercing mapper to not be useless about unions
* update S3V2WriteTest

## Fixing behavior for iceberg

* update the AirbyteValue -> IcebergRecord to handle the new AirbyteValues
* update the mapper pipeline (same as avro/parquet)
* add an expected record mapper, to account for iceberg not having a TimeTZ type
* update test disablings - enable testBasicTypes, disable testDedup

also fixes an edge case behavior that isn't covered by the new tests (this is why the pipeline swapped from SchemalessTypesToStringType to noop schema mapper). Submitted https://github.com/airbytehq/airbyte-internal-issues/issues/11218 to catch this. 

## Update unit tests

* Move a bunch of timestamp-parsing tests from TimeStringToIntegerTest -> AirbyteValueDeepCoercingMapperTest
* update AirbyteValueIdentityMapperTest - identity mapper no longer nulls anything
* misc compiler fixes

### CODE REVIEW

File: airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/IcebergParquetPipelineFactory.kt [hunk 1]
```diff
@@ -20,10 +22,10 @@ class IcebergParquetPipelineFactory : MapperPipelineFactory {
         MapperPipeline(
             stream.schema,
             listOf(
-                // TODO not sure why base parquet was doing this as a noop
-                SchemalessTypesToStringType() to SchemalessValuesToJsonString(),
-                AirbyteSchemaNoopMapper() to NullOutOfRangeIntegers(),
+                SchemalessTypesToStringType to SchemalessValuesToJsonString(),
```
Comment (edgao) [inline airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/IcebergParquetPipelineFactory.kt:25]: open thread - should iceberg also include the "fail on unknown type" mapper that avro/parquet have https://airbytehq-team.slack.com/archives/C03AS1GAQV6/p1734968675676689?thread_ts=1734747025.083309&cid=C03AS1GAQV6

(this is causing the test failure)

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt [hunk 2]
```diff
@@ -5,202 +5,42 @@
...
-    }
-
-    private fun matchesPermissively(schema: AirbyteType, json: JsonNode): Boolean {
-        return try {
-            convert(json, schema) !is UnknownValue
-        } catch (t: Throwable) {
-            false
+            JsonNodeType.POJO,
+            JsonNodeType.MISSING,
+            JsonNodeType.BINARY ->
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/json/JsonToAirbyteValue.kt:38]: It will be interesting to see whether this breaks anything in the wild!

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import java.time.format.DateTimeFormatter
+
+/**
+ * A mapper which coerces ALL values against the schema. This mapper MUST NOT be called after any
+ * mapper that returns non-native-JSON types (date, timestamp, etc.), or any mapper that causes the
+ * values to become misaligned with the schema (e.g. [AirbyteSchemaNoopMapper] +
+ * [SchemalessValuesToJsonString]).
+ *
+ * If this mapper is included in a [MapperPipeline], it SHOULD be preceded by a [MergeUnions]
+ * mapper. Not including this mapper may result in strange behavior when coercing union types.
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt:26]: So if I'm following so far
Initial marshaling -> just turns the json nodes into their equivalent airbyte values, w/o respecting the schema
this mapper forces everything to align with the schema

so if i wanted to use this to aggressively filter in parquet

```
Marshal -> [Pipeline: MapTimestamps -> SchemalessToStrings -> MergeUnions -> Coerce -> ToDisjointRecord] -> ToAvro
```

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    ): Pair<AirbyteValue, Context> =
+        requireType<StringValue>(value, TimeTypeWithTimezone, context) {
+            val ot =
+                try {
+                    OffsetTime.parse(it.value, TIME_FORMATTER)
+                } catch (e: Exception) {
+                    LocalTime.parse(it.value, TIME_FORMATTER).atOffset(ZoneOffset.UTC)
+                }
+            TimeWithTimezoneValue(ot) to context
+        }
```
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt:130]: except that this requires that I be a string coming in. So i would have to put `ToTimestamps` (and `Schemaless`) *after* this. 

```Marshal -> [Pipeline: MergeUnions -> Coerce -> MapTimestamps -> SchemalessToStrings -> 
 ToDisjointRecord] -> ToAvro```

Hm... having the extra types would make this more flexible, yes? B/c they would be validated here. then i could do `MapTimestamps -> SchemalessToStrings -> MergeUnions -> ToDisjointRecord -> Coerce(ie validate)`, which makes more sense. Like, first map the types, then resolve the unions, then change how unions are represented, finally validate that all the values match their final types, then convert?

File: airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt [hunk 5]
```diff
@@ -140,7 +141,7 @@ class S3V2WriteTestCsvGzip :
 class S3V2WriteTestAvroUncompressed :
     S3V2WriteTest(
         S3V2TestUtils.AVRO_UNCOMPRESSED_CONFIG_PATH,
-        TODO(),
+        NoopExpectedRecordMapper,
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt:144]: It feels confusing that the stricter destinations use Noop, and the less strict use Uncoerced

**Comment by johnny-schmidt** [comment]
This looks right to me. I feel like we need to whiteboard out all the type/value issues after break and try to simplify them.

**Review by johnny-schmidt — APPROVED** [review]

File: airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/IcebergParquetPipelineFactory.kt (hunk 1 above)
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/data/iceberg/parquet/IcebergParquetPipelineFactory.kt]: Rrrrright. It would make the most sense to fail on Unknown in the coerce mapper, since the point of that is to validate.

But we can't because Avro is using Unknown as a sentinel type for NullType.

But... if we had an explicit NullType... 🙃

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt (hunk 3 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt:26]: sounds right? not sure what you're putting into MapTimestamps, but if that's operating on raw StringValue, then yes

(and putting SchemalessToString before the Coerce mapper means that we need to also modify the schema to be StringType, unless we have some way to tell ToDisjointRecord that it was originally an ObjectWithEmptySchema/etc, though that's the thing we're trying to solve in slack )

File: airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt (hunk 4 above)
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/data/AirbyteValueDeepCoercingMapper.kt:130]: hm. maybe a thing we should pick up once we're back from holidays, I think I'm mostly with you, but confused about putting MapTimestamps first. I'm thinking of Coerce as The Thing That Maps to Real Types, so anything that wants to operate on timestamps should come after Coerce

(but yeah, putting the schemalesstostring + todisjointrecord before Coerce is pretty clearly superior)

File: airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt (hunk 5 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-s3-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/s3_v2/S3V2WriteTest.kt:144]: this is a thing I'll probably change for https://github.com/airbytehq/airbyte-internal-issues/issues/11218. I think this is only b/c the avro/parquet dumpers are implicitly coercing back to the expected schema - I want to switch that to be more like how iceberg does it, i.e.
* dumper maps naively from AvroValue -> AirbyteValue (so e.g. it would read timestamps as IntegerValue)
* expected record mapper maps expected record AirbyteValue (so e.g. convert TimestampTZValue -> IntegerValue)

File: airbyte-integrations/connectors/destination-iceberg-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergV2WriteTest.kt [hunk 6]
```diff
@@ -80,6 +80,12 @@ abstract class IcebergV2WriteTest(
     override fun testDedupChangeCursor() {
         super.testDedupChangeCursor()
     }
+
+    @Test
+    @Disabled("https://github.com/airbytehq/airbyte-internal-issues/issues/11221")
```
Comment (johnny-schmidt) [inline airbyte-integrations/connectors/destination-iceberg-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergV2WriteTest.kt:85]: I added a `failOnUnknownTypes` flag that you can use to toggle whether to expect failure in this case

Comment (johnny-schmidt) [reply airbyte-integrations/connectors/destination-iceberg-v2/src/test-integration/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergV2WriteTest.kt:85]: Er your issue is more subtle https://airbytehq-team.slack.com/archives/C03AS1GAQV6/p1734968675676689?thread_ts=1734747025.083309&cid=C03AS1GAQV6

sorry my brain forgot everything over break who am i what is a "data"?
